src/buffer/out/textBuffer.cpp

```diff
@@ -2353,3 +2354,86 @@ void TextBuffer::CopyHyperlinkMaps(const TextBuffer& other)
+// Arguments:
+// - The firstRow to start searching from
+// - The lastRow to search
+// Return value:
+// - A vector containing 3-element tuples where
+//      tuple<0>: ID of the pattern
+//      tuple<1>: start coordinate of the pattern
+//      tuple<2>: end coordinate of the pattern
+const std::vector<std::tuple<size_t, COORD, COORD>> TextBuffer::UpdatePatterns(const size_t firstRow, const size_t lastRow) const
+{
+    std::vector<std::tuple<size_t, COORD, COORD>> result;
+
+    std::wstring concatAll;
+    const auto rowSize = GetRowByOffset(0).size();
+    concatAll.reserve(rowSize * (lastRow - firstRow + 1));
+
+    // to deal with text that spans multiple lines, we will first concatenate
+    // all the text into one string and find the pattern in that string
+    // later, we convert the locations of the patterns back into (row, col) coords
+    for (auto i = firstRow; i <= lastRow; ++i)
+    {
+        auto row = GetRowByOffset(i);
+        concatAll += row.GetCharRow().GetText();
+    }
+
+    // convert the string into something the regex iterator can work with
+    const auto constAll = til::u16u8(concatAll);
+
+    // for each pattern we know of, iterate through the string
+    for (auto idAndPattern : _IdsAndPatterns)
+    {
+        std::regex regexObj{ idAndPattern.second };
+
+        // search through the run with our regex object
+        auto words_begin = std::sregex_iterator(constAll.begin(), constAll.end(), regexObj);
+        auto words_end = std::sregex_iterator();
+
+        size_t lenUpToThis = 0;
+        for (auto i = words_begin; i != words_end; ++i)
+        {
+            // record the locations and convert them back to coords
+            // when we find a match, the prefix is text that is between this
+            // match and the previous match, and the suffix is the text that
+            // it between this match and the next match
+            // we will use that, along with the size of the match, to determine
+            // the locations
+            const auto prefixSize = i->prefix().str().size();
+            const auto start = lenUpToThis + prefixSize;
+            const auto end = start + i->str().size();
+            lenUpToThis = end;
+
+            // store the locations as (col, row) coordinates
+            // NOTE: these are VIEWPORT coordinates, not buffer coordinates
+            // Keeping these as viewport coordinates for now because its the renderer
+            // that actually uses these coordinates and the renderer works in viewport coords
+            const auto startRow = gsl::narrow<SHORT>(start / rowSize);
+            const auto startCol = gsl::narrow<SHORT>(start % rowSize);
+            const auto endRow = gsl::narrow<SHORT>(end / rowSize);
+            const auto endCol = gsl::narrow<SHORT>(end % rowSize);
+            const COORD startCoord{ startCol, startRow };
+            const COORD endCoord{ endCol, endRow };

```

Comment:
I am pretty sure this level of math is also done in Find/Selection and it might be using TIL or one of the utilities/types classes to help with it. Consider deduplicating with those.

Reply:
Hm I can't seem to find the similar math - those parts of the code seem to begin with coordinates already (buffer or viewport) whereas here, since we concatenate all the text together, before searching, we need to calculate the coordinates from the concatenated string